### PR TITLE
Add coordinate conversion helper functions

### PR DESCRIPTION
**Pull request by hongyuchen1030**
Add 3D and 2D coordinates conversion functions into the uxarray/helpers.py: convert_node_xyz_to_latlon_rad; convert_node_latlon_rad_to_xyz ; normalize_in_place

In the following implementation, we will need to calculate the geometry using the 3D Cartesian Coordinates since it's more commonly used in math. And meanwhile, our users will get more used to the longitude and latitude coordinates for each point on the sphere. Therefore, we will need the coordinates conversion tools. 

And the normalize_in_place function is used to ensure all of our calculation is based on a unit sphere.

### CODE REVIEW

**Comment by rajeeja** [comment]
please resolve conflicts and remove Draft from PR title. Changes look fine, just need small unittest for the newly added functions.

**Comment by rajeeja** [comment]
pre-commit, documentation in docs folder and test in test folder needed.

**Comment by hongyuchen1030** [comment]
> pre-commit, documentation in docs folder and test in test folder needed.

Thank you for your feedback. I have merged the main branch to the PR. And I am adding unittest cases for the the helper functions now

**Comment by rajeeja** [comment]
> > pre-commit, documentation in docs folder and test in test folder needed.
> 
> Thank you for your feedback. I have merged the main branch to the PR. And I am adding unittest cases for the the helper functions now

docs?. run` make html `in docs folder.

**Comment by rajeeja** [comment]
@hongyuchen1030 thanks for the changes, please look at docs in the above. https://uxarray--90.org.readthedocs.build/en/90/

I don't see the user api change reflect: https://uxarray--90.org.readthedocs.build/en/90/user_api/index.html, 

Good work, now all the checks are passing.

**Comment by anissa111** [comment]
Hey @hongyuchen1030, I've made a PR on your fork that addresses your documentation troubles! https://github.com/hongyuchen1030/uxarray/pull/2

**Review by rajeeja — CHANGES_REQUESTED** [review]
I think a new test_edge for edge class would help understand this class and options provided.

**Comment by hongyuchen1030** [comment]
> I think a new test_edge for edge class would help understand this class and options provided.

I have added the test cases for the edge class in a new file called test_edge.py in the./test directory

**Review by rajeeja — APPROVED** [review]

**Comment by rajeeja** [comment]
fix precommit

**Review by erogluorhan — COMMENTED** [review]
Thank you very much for the effort! 

I would like to share a few thoughts to point things that I believe are significant about the design decisions, documentation, and implementation. 

I am happy to share here, in a dedicated development meeting, or during the Friday's discussion meeting at the least. Please let me know if you have a preference.

**Comment by hongyuchen1030** [comment]
> Thank you very much for the effort!
> 
> I would like to share a few thoughts to point things that I believe are significant about the design decisions, documentation, and implementation.
> 
> I am happy to share here, in a dedicated developers' meeting, or during the Friday's discussion meeting at the least. Please let me know if you have a preference.

Thank you so much for your information! I wonder if you can share it here? Thanks.

**Comment by erogluorhan** [comment]
> Thank you so much for your information! I wonder if you can share it here? Thanks.

Sure! 

- This PR is introducing a new class "Edge", which is a design change. However:
   - There is not an existing issue in this repo that this PR is linked to and trying to address. 
   - Such a class was neither proposed in the [draft UXarray API](https://github.com/UXARRAY/uxarray/blob/main/docs/user_api/uxarray_api.md) so far, nor it was discussed through any meeting AFAIK. 
   - It is not being documented in detail in this PR either (its objectives, future direction, i.e. possible future functions, where the current implemented function will be used, etc). Nor this PR is adding it into the draft API.
      - So, it is hard to foresee the need for this class (and a design change as a result) as well as its future, and it is difficult to review the PR.
         -  I'd suggest starting with describing the need for such a class and its future in discussion meetings and/or GitHub issues/discussions.
- Addition of helper functions for coordinate conversion and normalization seems to be irrelevant from the scope of this PR. They need to be a separate PR I believe.

**Comment by hongyuchen1030** [comment]
> > Thank you so much for your information! I wonder if you can share it here? Thanks.
> 
> Sure!
> 
> * This PR is introducing a new class "Edge", which is a design change. However:
>   
>   * There is not an existing issue in this repo that this PR is linked to and trying to address.
>   * Such a class was neither proposed in the [draft UXarray API](https://github.com/UXARRAY/uxarray/blob/main/docs/user_api/uxarray_api.md) so far, nor it was discussed through any meeting AFAIK.
>   * It is not being documented in detail in this PR either (its objectives, future direction, i.e. possible future functions, where the current implemented function will be used, etc). Nor this PR is adding it into the draft API.
>     
>     * So, it is hard to foresee the need for this class (and a design change as a result) as well as its future, and it is difficult to review the PR.
>       
>       * I'd suggest starting with describing the need for such a class and its future in discussion meetings and/or GitHub issues/discussions.
> * Addition of helper functions for coordinate conversion and normalization seems to be irrelevant from the scope of this PR. They need to be a separate PR I believe.
> * EDIT: A few more but little observations: The PR is still in "Draft" (i.e. not-ready-for-review) status. The title might be a bit more descriptive,

Thank you for your information! 

- About the "Edge" Class, this is a helper Object for `build_edge_face_connectivity`. Since all edges are undirected, I need to use a custom `__hash__`  and `__eq__` function to define the equality of the two edges.  My intention is that this is an internal helper class and I didn't plan for it to become a user API. I also asked for advice about a better solution. But so far, I am unable to find a more concise way to implement an undirected edge. But I am also open to adopting any better suggestion to make this function works.

- I have proposed this idea and talked about the code architectures during the meetings. And during the meeting, we think it's okay to make it an independent class in the uxarray. But any ideas are welcome to make the project structure better

- This PR is the first PR of the `buildlatlon_bounds` function, which is a very complicated and bulk function. Helper functions for coordinate conversion and normalization, along with the `Edge`, lays the foundation of all following functions PR, that's also the title for this PR. I consider the total code amounts so far still reasonable. But I am happy to break these functions down into separate PR if needed.

- As for the status of this PR, I wonder if that's because I named my branch with "DRAFT_Classes_n_HelperFunctions_Chen"? Or is there anything I can do to make this PR out of the "draft" status? Thank you very much!

**Comment by erogluorhan** [comment]
> * As for the status of this PR, I wonder if that's because I named my branch with "DRAFT_Classes_n_HelperFunctions_Chen"? Or is there anything I can do to make this PR out of the "draft" status? Thank you very much!

Oh, never mind; There is no problem with the status of this PR! It is my mistake!

**Comment by erogluorhan** [comment]
> * About the "Edge" Class, this is a helper Object for `build_edge_face_connectivity`. Since all edges are undirected, I need to use a custom `__hash__`  and `__eq__` function to define the equality of the two edges.  My intention is that this is an internal helper class and I didn't plan for it to become a user API. I also asked for advice about a better solution. But so far, I am unable to find a more concise way to implement an undirected edge. But I am also open to adopting any better suggestion to make this function works.

Thank you very much for the clarification; it helps! I have a few follow-up points on this:

- The above clear description of the `Edge` class was what I was suggesting previously. At the least, creating an issue that explains the need for such a helper code entity (either a class or something else) and linking to that issue from this PR would always help the readers understand what this PR is addressing and so keep track of tasks easily. Could you please create an issue.
- I am personally fine now with such an internal class for this purpose. @rajeeja @anissa111 @michaelavs please provide your inputs as well, if any, in the issue @hongyuchen1030 will create.
- If the final decision is still to go with such an internal class, I'd request you to:
   - Rename `edge.py` to `_edge.py`.
   - Rename `class Edge` to `class _Edge`
   - Rename `def Edge.get_nodes()` to `def _Edge._get_nodes()`
   to consistently reflect it is for internal use only.

**Review by anissa111 — CHANGES_REQUESTED** [review]
I agree with @erogluorhan's comments above and have a few more thoughts to add. Also, I think some of us (@erogluorhan, @rajeeja, @hongyuchen1030 and me maybe?) should sit down at a dedicated meeting together to fast track changes to the PR. 

I am also not certain that the new Edge class introduced here is necessary. I think this should probably be a discussion point at the next Friday meeting.

 ---

> As for the status of this PR, I wonder if that's because I named my branch with "DRAFT_Classes_n_HelperFunctions_Chen"? Or is there anything I can do to make this PR out of the "draft" status? Thank you very much!

It looks to be out of draft status for me! The naming of a branch shouldn't have any effect on the PR status. @erogluorhan I think you might be experiencing some bug on your end with github.

---

> This PR is the first PR of the buildlatlon_bounds function, which is a very complicated and bulk function. Helper functions for coordinate conversion and normalization, along with the Edge, lays the foundation of all following functions PR, that's also the title for this PR. I consider the total code amounts so far still reasonable. But I am happy to break these functions down into separate PR if needed.

Ah! This answers another question that I was going to ask in this PR. I think I would still like some clarification on the overall direction of the planned PRs. I think it's possible that I'm misunderstanding the purpose of some parts of this PR because I'm not understanding your future plans for it.

Let me know if there's anything I can do to clarify my review or comments!

**Comment by erogluorhan** [comment]
> * I have proposed this idea and talked about the code architectures during the meetings. And during the meeting, we think it's okay to make it an independent class in the uxarray. But any ideas are welcome to make the project structure better

Sorry if I missed them; I checked through the meeting notes that I couldn't attend, but I hadn't seen some details. Maybe this is showing us a need for forming a UXarray infrastructure working group or considering some alternating bi-weekly developers' meeting, which could help all of us discuss architectural dev etc? (The current UXarray meeting is only an-hour log bi-weekly, so it is not easy to discuss all the technicality there I think). Just a thought.

**Comment by erogluorhan** [comment]
> * This PR is the first PR of the `buildlatlon_bounds` function, which is a very complicated and bulk function. Helper functions for coordinate conversion and normalization, along with the `Edge`, lays the foundation of all following functions PR, that's also the title for this PR. I consider the total code amounts so far still reasonable. But I am happy to break these functions down into separate PR if needed.

This is also relatable to my suggestion with the first bullet-point. The clarification here works very well for me, and it should absolutely go into a dedicated GitHub issue that describes the problem and the needs in detail (Without this explanation here, it was very difficult to understand what actual needs this PR was trying to address). Then, this PR could just link to that issue. This would also work great for future reference, even a long while after this work gets merged.

File: docs/internal_api/index.rst
Comment (hongyuchen1030) [reply docs/internal_api/index.rst:32]: Yes, this entire class is for internal use only so far

**Comment by hongyuchen1030** [comment]
> > * About the "Edge" Class, this is a helper Object for `build_edge_face_connectivity`. Since all edges are undirected, I need to use a custom `__hash__`  and `__eq__` function to define the equality of the two edges.  My intention is that this is an internal helper class and I didn't plan for it to become a user API. I also asked for advice about a better solution. But so far, I am unable to find a more concise way to implement an undirected edge. But I am also open to adopting any better suggestion to make this function works.
> 
> Thank you very much for the clarification; it helps! I have a few follow-up points on this:
> 
> * The above clear description of the `Edge` class was what I was suggesting previously. At the least, creating an issue that explains the need for such a helper code entity (either a class or something else) and linking to that issue from this PR would always help the readers understand what this PR is addressing and so keep track of tasks easily. Could you please create an issue.
> * I am personally fine now with such an internal class for this purpose. @rajeeja @anissa111 @michaelavs please provide your inputs as well, if any, in the issue @hongyuchen1030 will create.
> * If the final decision is still to go with such an internal class, I'd request you to:
>   
>   * Rename `edge.py` to `_edge.py`.
>   * Rename `class Edge` to `class _Edge`
>   * Rename `def Edge.get_nodes()` to `def _Edge._get_nodes()`
>     to consistently reflect it is for internal use only.

Thank you for your suggestion. I have created the issue  #111

File: docs/internal_api/index.rst
Comment (hongyuchen1030) [reply docs/internal_api/index.rst]: Thank you for your suggestion.
I have fixed this in my latest commit

File: uxarray/helpers.py
Comment (hongyuchen1030) [reply uxarray/helpers.py]: Thank you for your suggestion.
I have fixed this in my latest commit

**Comment by hongyuchen1030** [comment]
> > * This PR is the first PR of the `buildlatlon_bounds` function, which is a very complicated and bulk function. Helper functions for coordinate conversion and normalization, along with the `Edge`, lays the foundation of all following functions PR, that's also the title for this PR. I consider the total code amounts so far still reasonable. But I am happy to break these functions down into separate PR if needed.
> 
> This is also relatable to my suggestion with the first bullet-point. The clarification here works very well for me, and it should absolutely go into a dedicated GitHub issue that describes the problem and the needs in detail (Without this explanation here, it was very difficult to understand what actual needs this PR was trying to address). Then, this PR could just link to that issue. This would also work great for future reference, even a long while after this work gets merged.

I have created the issue #112

**Comment by erogluorhan** [comment]
> Thank you for your suggestion. I have created the issue #111

Thanks! Please address the requests in the previous comment to reflect Edge is internal API-only.

**Comment by hongyuchen1030** [comment]
I have already changed the `edge` class into the internal API. And this sounds like a reasonable change so far. But the documentation population setting might need a further check since I rename the file.

**Comment by hongyuchen1030** [comment]
I have removed the `Edge` class and will replace it with the `frozenset` object in the following implementations and PRs

**Review by hongyuchen1030 — COMMENTED** [review]
After discussion with @erogluorhan, I have decided to replace the `Edge` class with the python `frozenset`

**Comment by hongyuchen1030** [comment]
@anissa111 I have updated the PR according to our meeting last Thursday:
1. The documentations format errors are fixed
2. Use the `np.linalg.norm` internally to replace the original manual calculation in the `normalize_in_place`. We still want to wrap the `np.linalg.norm` up because we can check for the input length. Also, it can make the usages look cleaner
3. Edit the PR description and add more details

**Review by erogluorhan — APPROVED** [review]
This looks great; approved! Could you please do the followings to make it look perfect:

1. Change the title of PR to reflect its work better. Maybe something like "Add coordinate conversion helper functions"
2. Sort the function names in the `docs/internal_api/index.rst` alphabetically?

File: uxarray/helpers.py
Comment (anissa111) [inline uxarray/helpers.py:499]:
```suggestion
    float list
        the result array of the unit 3D coordinates [x, y, z] vector where :math:`x^2 + y^2 + z^2 = 1`
```

File: uxarray/helpers.py
Comment (anissa111) [inline uxarray/helpers.py:572]:
```suggestion
        the result unit vector [x, y, z] where :math:`x^2 + y^2 + z^2 = 1`
```

**Review by anissa111 — APPROVED** [review]
I agree with Orhan's review above and also have added suggestions below that will help render the equations in your docstrings.

Great job!

**Comment by hongyuchen1030** [comment]
@anissa111 @erogluorhan 
Thank you for your information
I have already updated the branches according to your suggestions.

**Pull request merged**
